**Описание проекта**  
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**  

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Задания**  

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Как выполнить задание?**  

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

## Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Обзор таблиц

In [2]:
for data in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(data) 

    request = pd.io.sql.read_sql(query, con = engine)
    display(f'Таблица - "{data}"')
    display(request.info())
    display(request.head())   

'Таблица - "books"'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

'Таблица - "authors"'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

'Таблица - "publishers"'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

'Таблица - "ratings"'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

'Таблица - "reviews"'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

* Все таблицы выглядят как указано в описании, формат данных указан корректно, пропусков по info не видно


## Выполнение задач
### Посчитайте, сколько книг вышло после 1 января 2000 года

In [3]:
first_task = '''
SELECT COUNT(book_id) AS count_book
FROM books
WHERE publication_date >= '2000-01-01';
'''
first_task = pd.io.sql.read_sql(first_task, con = engine)
first_task

count_book
0         821

**Ответ:**  
У нас имеются данные, о том что после 1 января 2000 года, вышло 821 книга.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
second_task = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT re.review_id) AS reviews_cnt, 
       AVG (CAST(r.rating AS integer)) AS avg_ratings
FROM books AS b
FULL JOIN reviews AS re ON re.book_id = b.book_id
FULL JOIN ratings AS r ON r.book_id = b.book_id
GROUP BY b.book_id,
         b.title
ORDER BY reviews_cnt DESC,
         avg_ratings DESC         
'''
second_task = pd.io.sql.read_sql(second_task, con = engine)
second_task

book_id                                              title  reviews_cnt  \
0        948                            Twilight (Twilight  #1)            7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
2        299  Harry Potter and the Chamber of Secrets (Harry...            6   
3        656                                     The Book Thief            6   
4        734                                   The Glass Castle            6   
..       ...                                                ...          ...   
995      191  Disney's Beauty and the Beast (A Little Golden...            0   
996      387                               Leonardo's Notebooks            0   
997      221                          Essential Tales and Poems            0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
999      808                            The Natural Way to Draw            0   

     avg_ratings  
0       3.662500  
1       4.414634  
2       4.287500  
3       4.264151  
4       4.206897  
..           ...  
995     4.000000  
996     4.000000  
997     4.000000  
998     3.666667  
999     3.000000  

[1000 rows x 4 columns]

**Ответ:**  
На все книги выведено количество обзоров и средняя оценка. Наибольшим количеством обзоров обладает первая книга "Сумерек".

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [5]:
third_task = '''
SELECT p.publisher AS publisher,
       COUNT(b.title) AS book_count
FROM publishers AS p
JOIN books AS b ON b.publisher_id = p.publisher_id
WHERE num_pages >= 50
GROUP BY publisher
ORDER BY book_count DESC;
'''
third_task = pd.io.sql.read_sql(third_task, con = engine)
third_task.head()

publisher  book_count
0             Penguin Books          42
1                   Vintage          31
2  Grand Central Publishing          25
3          Penguin Classics          24
4          Ballantine Books          19

**Ответ:**  
Наибольшее число книг толще 50 страниц выпустил издатель Penguin Books.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [6]:
fourth_task = '''
SELECT query.author,
       ROUND(AVG(query.avg_rating),2) AS avg_rate
FROM
        (SELECT a.author AS author,
                AVG(r.rating) AS avg_rating,
                COUNT(r.rating_id) AS count_rating
        FROM books AS b
        LEFT JOIN authors AS a ON b.author_id = a.author_id
        LEFT JOIN ratings AS r ON b.book_id = r.book_id
        GROUP BY b.book_id, a.author_id
        HAVING COUNT(r.rating_id) >=50 
        ) AS query
GROUP BY query.author
ORDER BY avg_rate DESC;
'''
fourth_task = pd.io.sql.read_sql(fourth_task, con = engine)
fourth_task.head()

author  avg_rate
0         J.K. Rowling/Mary GrandPré      4.28
1  Markus Zusak/Cao Xuân Việt Khương      4.26
2                     J.R.R. Tolkien      4.26
3                  Louisa May Alcott      4.19
4                       Rick Riordan      4.08

**Ответ:**  
Автор с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками, становится J.K. Rowling/Mary GrandPré	

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [7]:
fifth_task = ''' 
SELECT ROUND(AVG(re.сount_reviews),2) as avg_reviews
FROM
    (SELECT username,
            COUNT(review_id) AS сount_reviews
    FROM reviews AS r
    WHERE username IN 
                (SELECT username
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48)
    GROUP BY username) AS re; 
''' 
fifth_task = pd.io.sql.read_sql(fifth_task, con = engine)
fifth_task

avg_reviews
0         24.0

**Ответ:**  
Среднее количество обзоров от пользователей, которые поставили больше 48 оценок составляет 24.